File: modules/vaine-face-detection/android/src/main/java/expo/modules/vainefacedetection/VaineFaceDetectionModule.kt
package expo.modules.vainefacedetection

import android.net.Uri
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.face.FaceLandmark
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Still-image face detection for capture alignment.
 *
 * ML Kit runs in accurate mode with landmarks only: contours, classification
 * (smiling / eyes-open), and tracking stay disabled so the module cannot
 * produce identity, demographic, or emotion signals. InputImage.fromFilePath
 * applies EXIF rotation; coordinates are reported in the upright space, and
 * the payload carries the upright pixel dimensions so JavaScript never has to
 * guess the coordinate space.
 */
class VaineFaceDetectionModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("VaineFaceDetection")

    AsyncFunction("detectFaces") { imageUri: String ->
      val context = appContext.reactContext
        ?: throw CodedException("ERR_VAINE_FACE_DETECTION", "React context is unavailable.", null)

      val image = InputImage.fromFilePath(context, Uri.parse(imageUri))
      val options = FaceDetectorOptions.Builder()
        .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
        .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
        .setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
        .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
        .setMinFaceSize(0.15f)
        .build()

      val detector = FaceDetection.getClient(options)
      try {
        val faces = Tasks.await(detector.process(image))
        val rotated = image.rotationDegrees == 90 || image.rotationDegrees == 270
        val uprightWidth = if (rotated) image.height else image.width
        val uprightHeight = if (rotated) image.width else image.height

        val landmarkTypes = mapOf(
          "leftEye" to FaceLandmark.LEFT_EYE,
          "rightEye" to FaceLandmark.RIGHT_EYE,
          "noseBase" to FaceLandmark.NOSE_BASE,
          "mouthLeft" to FaceLandmark.MOUTH_LEFT,
          "mouthRight" to FaceLandmark.MOUTH_RIGHT,
          "mouthBottom" to FaceLandmark.MOUTH_BOTTOM,
        )

        mapOf(
          "imageWidth" to uprightWidth,
          "imageHeight" to uprightHeight,
          "faces" to faces.map { face ->
            val landmarks = mutableMapOf<String, Map<String, Float>>()
            for ((key, type) in landmarkTypes) {
              face.getLandmark(type)?.let { landmark ->
                landmarks[key] = mapOf("x" to landmark.position.x, "y" to landmark.position.y)
              }
            }
            mapOf(
              "frame" to mapOf(
                "x" to face.boundingBox.left,
                "y" to face.boundingBox.top,
                "width" to face.boundingBox.width(),
                "height" to face.boundingBox.height(),
              ),
              "landmarks" to landmarks,
            )
          },
        )
      } finally {
        detector.close()
      }
    }
  }
}

File: modules/vaine-face-detection/ios/VaineFaceDetectionModule.swift
import ExpoModulesCore
import Foundation
import MLKitFaceDetection
import MLKitVision
import UIKit

/**
 Still-image face detection for capture alignment.

 ML Kit is configured for accurate single-shot analysis with landmarks only:
 contours, classification (smiling / eyes-open), and tracking stay disabled so
 the module cannot produce identity, demographic, or emotion signals. The
 image is orientation-normalized before detection and the payload reports the
 exact pixel dimensions the detector measured in, so JavaScript never has to
 guess the coordinate space.
 */
public class VaineFaceDetectionModule: Module {
  public func definition() -> ModuleDefinition {
    Name("VaineFaceDetection")

    AsyncFunction("detectFaces") { (imageUri: String) -> [String: Any] in
      guard let url = URL(string: imageUri), url.isFileURL,
            let originalImage = UIImage(contentsOfFile: url.path) else {
        throw ImageLoadException(imageUri)
      }

      let image = Self.normalizedOrientation(originalImage)
      let visionImage = VisionImage(image: image)
      visionImage.orientation = .up

      let options = FaceDetectorOptions()
      options.performanceMode = .accurate
      options.landmarkMode = .all
      options.contourMode = .none
      options.classificationMode = .none
      options.isTrackingEnabled = false
      options.minFaceSize = 0.15

      let detector = FaceDetector.faceDetector(options: options)
      let faces = try detector.results(in: visionImage)

      let pixelWidth = image.size.width * image.scale
      let pixelHeight = image.size.height * image.scale

      let landmarkTypes: [(String, FaceLandmarkType)] = [
        ("leftEye", .leftEye),
        ("rightEye", .rightEye),
        ("noseBase", .noseBase),
        ("mouthLeft", .mouthLeft),
        ("mouthRight", .mouthRight),
        ("mouthBottom", .mouthBottom),
      ]

      return [
        "imageWidth": pixelWidth,
        "imageHeight": pixelHeight,
        "faces": faces.map { face -> [String: Any] in
          var landmarks: [String: Any] = [:]
          for (key, type) in landmarkTypes {
            if let landmark = face.landmark(ofType: type) {
              landmarks[key] = ["x": landmark.position.x, "y": landmark.position.y]
            }
          }
          return [
            "frame": [
              "x": face.frame.origin.x,
              "y": face.frame.origin.y,
              "width": face.frame.size.width,
              "height": face.frame.size.height,
            ],
            "landmarks": landmarks,
          ]
        },
      ]
    }
  }

  /// Re-renders the image upright so detector coordinates and the reported
  /// dimensions share one EXIF-independent pixel space.
  private static func normalizedOrientation(_ image: UIImage) -> UIImage {
    if image.imageOrientation == .up {
      return image
    }
    UIGraphicsBeginImageContextWithOptions(image.size, false, image.scale)
    defer { UIGraphicsEndImageContext() }
    image.draw(in: CGRect(origin: .zero, size: image.size))
    return UIGraphicsGetImageFromCurrentImageContext() ?? image
  }
}

internal final class ImageLoadException: GenericException<String> {
  override var reason: String {
    "Could not load an image from the local file URI: \(param)"
  }
}
